src/Streamly/Internal/Unicode/Utf8/Type.hs

```diff
@@ -0,0 +1,501 @@
+-- 'Utf8'.  Subject to fusion.
+{-# NOINLINE append #-}
+append :: Utf8 -> Utf8 -> Utf8
+append (Utf8 a) (Utf8 b) = Utf8 $ unsafePerformIO $ Array.spliceExp a b
+
+{-
+{-# RULES
+"TEXT append -> fused" [~1] forall t1 t2.
+    append t1 t2 = unstream (S.append (stream t1) (stream t2))
+"TEXT append -> unfused" [1] forall t1 t2.
+    unstream (S.append (stream t1) (stream t2)) = append t1 t2
+ #-}
+-}
+
+-- | /O(1)/ Returns the first character of a 'Utf8', or 'Nothing' if empty.
+-- Subject to fusion.
+{-# INLINE head #-}
+head :: Utf8 -> Maybe Char
+head = unsafePerformIO . Stream.head . stream
+
+-- XXX INLINE_NORMAL cause text. Need to find the reason.
+-- | /O(1)/ Returns the first character and rest of a 'Utf8', or 'Nothing' if
+-- empty. Subject to fusion.
+{-# INLINE_NORMAL uncons #-}
+uncons :: Utf8 -> Maybe (Char, Utf8)
+uncons = fmap (second unstream) . unsafePerformIO . Stream.uncons . stream
+
+-- | /O(1)/ Returns the last character of a 'Utf8', or 'Nothing' if empty.
+-- Subject to fusion.
+{-# INLINE_NORMAL last #-}
+last :: Utf8 -> Char
+last = undefined
+
+{-
+{-# RULES
+"TEXT last -> fused" [~1] forall t.
+    last t = S.last (stream t)
+"TEXT last -> unfused" [1] forall t.
+    S.last (stream t) = last t
+  #-}
+-}
+
+-- | /O(1)/ Returns all characters after the head of a 'Utf8', or 'Nothing' if
+-- empty.  Subject to fusion.
+{-# INLINE_NORMAL tail #-}
+tail :: Utf8 -> Maybe Utf8
+tail = fmap unstream . unsafePerformIO . Stream.tail . stream
+
+{-
+{-# RULES
+"TEXT tail -> fused" [~1] forall t.
+    tail t = unstream (S.tail (stream t))
+"TEXT tail -> unfused" [1] forall t.
+    unstream (S.tail (stream t)) = tail t
+ #-}
+-}
+
+-- | /O(1)/ Returns all but the last character of a 'Utf8', or 'Nothing' if
+-- empty.  Subject to fusion.
+{-# INLINE_NORMAL init #-}
+init :: Utf8 -> Maybe Utf8
+init = fmap unstream . unsafePerformIO . Stream.init . stream
+
+{-
+{-# RULES
+"TEXT init -> fused" [~1] forall t.
+    init t = unstream (S.init (stream t))
+"TEXT init -> unfused" [1] forall t.
+    unstream (S.init (stream t)) = init t
+ #-}
+-}
+
+-- | /O(1)/ Returns all but the last character and the last character of a
+-- 'Utf8', or 'Nothing' if empty.
+{-# INLINE unsnoc #-}
+unsnoc :: Utf8 -> Maybe (Utf8, Char)
+unsnoc = undefined
+
+-- | /O(1)/ Tests whether a 'Utf8' is empty or not.  Subject to fusion.
+{-# INLINE null #-}
+null :: Utf8 -> Bool
+null = Array.null . toArray
+
+{-
+{-# RULES
+"TEXT null -> fused" [~1] forall t.
+    null t = S.null (stream t)
+"TEXT null -> unfused" [1] forall t.
+    S.null (stream t) = null t
+ #-}
+-}
+
+-- | /O(1)/ Tests whether a 'Utf8' contains exactly one character.  Subject to
+-- fusion.
+{-# INLINE isSingleton #-}
+isSingleton :: Utf8 -> Bool
+isSingleton = undefined
+
+-- | /O(n)/ Returns the number of characters in a 'Utf8'.  Subject to fusion.
+{-# INLINE length #-}
+length :: Utf8 -> Int
+length = unsafePerformIO . Stream.length . stream
```

Comment:
We could possibly determine the length faster by using a custom routine that counts the starting chars from the utf8 encoded bytes without decoding the chars.